include/gaussian_samplers.h

```diff
@@ -0,0 +1,365 @@
+    //RNGType rng(seed);
+    RNGType &rng2 = var.rng;
+    if(a_i>0.00000001 && std::sqrt(bef.squared_length()) >= (2.0/std::sqrt(2.0*a_i))){
+        boost::normal_distribution<> rdist(0,1);
+        Point a = -1.0*lower;
+        Point b = (1.0/std::sqrt(bef.squared_length()))*bef;
+        Point z = (a.dot(b)*b)+lower;
+        NT low_bd = (lower[0]-z[0])/b[0];
+        NT up_bd = (upper[0]-z[0])/b[0];
+        while(true){
+            r = rdist(rng2);
+            r = r/std::sqrt(2.0*a_i);
+            if(r>=low_bd && r<=up_bd){
+                break;
+            }
+        }
+        p=(r*b)+z;
+
+    }else{
+        boost::random::uniform_real_distribution<> urdist(0,1);
+        NT M=get_max(lower, upper, a_i);
+        while(true){
+            r=urdist(rng2);
+            Point pef = r*upper;
+            p = ((1.0-r)*lower)+ pef;
+            r_val = M*urdist(var.rng);
+            fn = eval_exp(p,a_i);
+            if(r_val<fn){
+                break;
+            }
+        }
+    }
+    return 1;
+}
+
+
+int rand_exp_range_coord(NT l, NT u, NT a_i, NT &dis, vars_g &var){
+    NT r, r_val, fn;
+    //unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    //RNGType rng(seed);
+    RNGType &rng2 = var.rng;
+    if(a_i>std::pow(10,-8.0) && u-l>=2.0/std::sqrt(2.0*a_i)){
+        boost::normal_distribution<> rdist(0,1);
+        while(true){
+            r = rdist(rng2);
+            r = r/std::sqrt(2.0*a_i);
+            if(r>=l && r<=u){
+                break;
+            }
+        }
+        dis=r;
+
+    }else{
+        boost::random::uniform_real_distribution<> urdist(0,1);
+        NT M = get_max_coord(l, u, a_i);
+        while(true){
+            r=urdist(rng2);
+            dis = (1.0-r)*l+r*u;
+            r_val = M*urdist(rng2);
+            fn = std::exp(-a_i*dis*dis);
+            if(r_val<fn){
+                break;
+            }
+        }
+    }
+
+    return 1;
+}
+
+
+template <class T>
+int gaussian_next_point(T &P,
+                        Point &p,   // a point to start
+                        Point &p_prev,
+                        int &coord_prev,
+                        int walk_len,
+                        NT a_i,
+                        std::vector<NT> &lamdas,
+                        vars_g &var)
+{
+    int n=var.n;
+    boost::random::uniform_int_distribution<> uidist(0,n-1);
+    //unsigned seed = std::chrono::system_clock::now().time_since_epoch().count();
+    //RNGType rng(seed);
+    RNGType &rng2 = var.rng;
+    NT ball_rad;
+
+    if(var.ball_walk) {
+        if(var.delta<0.0) {
+            ball_rad = 4.0 * var.che_rad / std::sqrt(std::max(1.0, a_i) * NT(n));
+        }else{
+            ball_rad=var.delta;
+        }
+        gaussian_ball_walk(p, P, a_i, ball_rad, var);
+    }else if(!var.coordinate){
+        gaussian_hit_and_run(p,P,a_i,var);
+    }else{
+        int rand_coord;
+        if(coord_prev==-1){
```

Comment:
this is going to be checked all the time and only enter once, could you propose a more efficient solution?

Reply:
The same holds for the polytope function line_intersect_coord, init is true only the first time. Should we override the existing functions for the first point of the coordinate directions hit-and-run both in class Polytope ( line_intersect_coord() ) and in samplers.h ( rand_point_generator() ) ?

Reply:
done. In addition, we have override gaussian_next_point to compute the first point for the CDHR. See related commits